### PR TITLE
PelotonMain

### PR DESCRIPTION
**Pull request by huangminchn**
issue #404 

A singleton object "PelotonMain" is created and it contains all the singleton objects in the system.

1. ThreadPool, NetworkManager, EpochManager and GCManager become member variable in PelotonMain. Whenever these 4 instances are needed, get them from the PelotonMain object instead of calling their static GetInstance function;
2. I didn't make TrafficCop as a member in PelotonMain because TrafficCop is not a global singleton instance (except for unit test); a new TrafficCop instance is created for each network connection;

TODO:
1. add other global singletons to PelotonMain;
2. modify test cases so that they retrieve global singletons from PelotonMain

### CODE REVIEW

**Comment by apavlo** [comment]
We need to have a longer discussion about this PR. We need to remove all of the singleton classes entirely and pass around a single `PelotonMain` object everywhere that needs it.

File: src/peloton_main/peloton_main.cpp [hunk 1]
```diff
@@ -1,16 +1,8 @@
-//===----------------------------------------------------------------------===//
 //
-//                         Peloton
+// Created by Min Huang on 8/28/17.
```
Comment (apavlo) [inline src/peloton_main/peloton_main.cpp]: Don't do this.

**Comment by apavlo** [comment]
@jarulraj @malin1993ml @pmenon @tomasic @hzxa21 @poojanilangekar -- I was thinking that maybe we want to merge this PR as a temporary solution to getting rid of all of the static singleton objects. It would be a first step to get rid of all of the global singleton calls, and then work to change the code to pass around `PelotonMain` handle correctly.

Thoughts?

File: src/include/peloton_main/peloton_main.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+#include "concurrency/epoch_manager.h"
+#include "gc/gc_manager.h"
+
+namespace peloton {
+
+//===--------------------------------------------------------------------===//
+// Global Setup and Teardown
+//===--------------------------------------------------------------------===//
+
+class PelotonMain {
```
Comment (seojungmin) [inline src/include/peloton_main/peloton_main.h]: Would there be any reason why we could not make this class be just "Peloton"? Peloton sounds better over PelotonMain.

Comment (pmenon) [reply src/include/peloton_main/peloton_main.h]: Or just `PelotonDB` or `Database`?

**Comment by poojanilangekar** [comment]
@apavlo Are we planning to merge this by the end of this week?

**Comment by apavlo** [comment]
@poojanilangekar No, I think we are going to re-evaluate our options with @pervazea and @tli2 after the new year.

**Comment by apavlo** [comment]
We're going to abandon this PR and revisit it next semester.

**Pull request closed without merging**